### PR TITLE
Provide the fetched data to fetch() consumers.

### PR DESCRIPTION
**Pull request by Ms2ger**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13597)

### CODE REVIEW

File: components/script/dom/request.rs [hunk 1]
```diff
@@ -662,20 +665,19 @@ impl BodyOperations for Request {
         self.BodyUsed()
     }
 
+    fn set_body_promise(&self, p: &Rc<Promise>, body_type: BodyType) {
+        self.body_used.set(true);
+        assert!(mem::replace(&mut *self.body_promise.borrow_mut(), Some((p.clone(), body_type))).is_none());
```
Comment (jdm) [inline components/script/dom/request.rs]: Same as later comment.

File: components/script/dom/response.rs [hunk 2]
```diff
@@ -194,18 +197,25 @@ impl BodyOperations for Response {
         self.BodyUsed()
     }
 
+    fn set_body_promise(&self, p: &Rc<Promise>, body_type: BodyType) {
+        self.body_used.set(true);
+        assert!(mem::replace(&mut *self.body_promise.borrow_mut(), Some((p.clone(), body_type))).is_none());
```
Comment (jdm) [inline components/script/dom/response.rs]: This is a complicated way of writing
```rust
assert!(self.body_promise.borrow().is_none());
*self.body_promise.borrow_mut() = Some(p.clone(), body_type);
```
which I would prefer.

File: components/script/body.rs [hunk 3]
```diff
@@ -49,41 +50,50 @@ pub fn consume_body<T: BodyOperations + Reflectable>(object: &T, body_type: Body
...
     // Steps 2-4
     // TODO: Body does not yet have a stream.
 
+    consume_body_with_promise(object, body_type, &promise);
+
+    promise
+}
+
+// https://fetch.spec.whatwg.org/#concept-body-consume-body
+#[allow(unrooted_must_root)]
```
Comment (jdm) [inline components/script/body.rs]: Is this necessary?

File: components/script/body.rs [hunk 4]
```diff
@@ -49,41 +50,50 @@ pub fn consume_body<T: BodyOperations + Reflectable>(object: &T, body_type: Body
...
-        Ok(results) => {
-            match results {
-                FetchedData::Text(s) => promise.resolve_native(cx, &USVString(s)),
-                FetchedData::Json(j) => promise.resolve_native(cx, &j),
-                FetchedData::BlobData(b) => promise.resolve_native(cx, &b),
-                FetchedData::FormData(f) => promise.resolve_native(cx, &f),
-            };
-        },
-        Err(err) => promise.reject_error(cx, err),
+    if let Some(body) = object.take_body() {
```
Comment (jdm) [inline components/script/body.rs]:
```rust
let body = match object.take_body() {
    Some(body) => body,
    None => return,
};
```

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
We probably need a compartment for the JSON case.

**Comment by larsbergstrom** [comment]
@bors-servo clean force retry 
 - infra buildbot restart

**Comment by Ms2ger** [comment]
Oh what the hell:
```
LLVM ERROR: IO failure on output stream.
```
twice on servo-linux2

@larsbergstrom ?

**Comment by larsbergstrom** [comment]
@Ms2ger disk full, please retry.

**Comment by larsbergstrom** [comment]
@bors-servo retry
- infra

**Pull request merged**